# <center> Topic 1. Exploratory data analysis with Pandas
## <center>Practice. Analyzing "Titanic" passengers

**Fill in the missing code ("You code here").**

In [1]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt

# Graphics in SVG format are more sharp and legible
%config InlineBackend.figure_format = 'svg'
pd.set_option("display.precision", 2)

**Read data into a Pandas DataFrame**

In [2]:
PATH_TO_DATASET = "data/titanic_train.csv"

In [3]:
data = pd.read_csv(PATH_TO_DATASET, index_col="PassengerId")

**First 5 rows**

In [4]:
data.head(5)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket   Fare Cabin Embarked  
PassengerId                                                        
1                1      0         A/5 21171   7.25   NaN        S  
2                1      0          PC 17599  71.28   C85        C  
3                0      0  STON/O2. 3101282   7.92   NaN        S  
4                1      0            113803  53.10  C123        S  
5                0      0            373450   8.05   NaN        S

In [5]:
data.describe()

Survived  Pclass     Age   SibSp   Parch    Fare
count    891.00  891.00  714.00  891.00  891.00  891.00
mean       0.38    2.31   29.70    0.52    0.38   32.20
std        0.49    0.84   14.53    1.10    0.81   49.69
min        0.00    1.00    0.42    0.00    0.00    0.00
25%        0.00    2.00   20.12    0.00    0.00    7.91
50%        0.00    3.00   28.00    0.00    0.00   14.45
75%        1.00    3.00   38.00    1.00    0.00   31.00
max        1.00    3.00   80.00    8.00    6.00  512.33

**Let's select those passengers who embarked in Cherbourg (Embarked=C) and paid > 200 pounds for their ticker (fare > 200).**

Make sure you understand how actually this construction works.

In [6]:
data[(data["Embarked"] == "C") & (data.Fare > 200)].head()

Survived  Pclass  \
PassengerId                     
119                 0       1   
259                 1       1   
300                 1       1   
312                 1       1   
378                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
259                                         Ward, Miss. Anna  female  35.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
378                                Widener, Mr. Harry Elkins    male  27.0   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  
PassengerId                                                            
119              0      1  PC 17558  247.52          B58 B60        C  
259              0      0  PC 17755  512.33              NaN        C  
300              0      1  PC 17558  247.52          B58 B60        C  
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C  
378              0      2    113503  211.50              C82        C

**We can sort these people by Fare in descending order.**

In [44]:
data[(data["Embarked"] == "C") & (data["Fare"] > 200)].sort_values(
    by="Fare", ascending=False
).head(10)

Survived  Pclass  \
PassengerId                     
259                 1       1   
680                 1       1   
738                 1       1   
312                 1       1   
743                 1       1   
119                 0       1   
300                 1       1   
381                 1       1   
717                 1       1   
558                 0       1   

                                                        Name     Sex   Age  \
PassengerId                                                                  
259                                         Ward, Miss. Anna  female  35.0   
680                       Cardeza, Mr. Thomas Drake Martinez    male  36.0   
738                                   Lesurer, Mr. Gustave J    male  35.0   
312                               Ryerson, Miss. Emily Borie  female  18.0   
743                    Ryerson, Miss. Susan Parker "Suzette"  female  21.0   
119                                 Baxter, Mr. Quigg Edmond    male  24.0   
300          Baxter, Mrs. James (Helene DeLaudeniere Chaput)  female  50.0   
381                                    Bidois, Miss. Rosalie  female  42.0   
717                            Endres, Miss. Caroline Louise  female  38.0   
558                                      Robbins, Mr. Victor    male   NaN   

             SibSp  Parch    Ticket    Fare            Cabin Embarked  \
PassengerId                                                             
259              0      0  PC 17755  512.33              NaN        C   
680              0      1  PC 17755  512.33      B51 B53 B55        C   
738              0      0  PC 17755  512.33             B101        C   
312              2      2  PC 17608  262.38  B57 B59 B63 B66        C   
743              2      2  PC 17608  262.38  B57 B59 B63 B66        C   
119              0      1  PC 17558  247.52          B58 B60        C   
300              0      1  PC 17558  247.52          B58 B60        C   
381              0      0  PC 17757  227.53              NaN        C   
717              0      0  PC 17757  227.53              C45        C   
558              0      0  PC 17757  227.53              NaN        C   

             Age_category  
PassengerId                
259                   2.0  
680                   2.0  
738                   2.0  
312                   1.0  
743                   1.0  
119                   1.0  
300                   2.0  
381                   2.0  
717                   2.0  
558                   NaN

**Let's create a new feature.**

In [8]:
def age_category(age):
    """
    < 30 -> 1
    >= 30, <55 -> 2
    >= 55 -> 3
    """
    if age < 30:
        return 1
    elif age < 55:
        return 2
    elif age >= 55:
        return 3

In [9]:
age_categories = [age_category(age) for age in data.Age]
data["Age_category"] = age_categories

**Another way is to do it with `apply`.**

In [10]:
data["Age_category"] = data["Age"].apply(age_category)

**1. How many men/women were there onboard?**
- 412 men and 479 women
- 314 men and 577 women
- 479 men and 412 women
- 577 men and 314 women

In [11]:
print(data[data["Sex"] == "male"].count())

Survived        577
Pclass          577
Name            577
Sex             577
Age             453
SibSp           577
Parch           577
Ticket          577
Fare            577
Cabin           107
Embarked        577
Age_category    453
dtype: int64


**2. Print the distribution of the `Pclass` feature. Then the same, but for men and women separately. How many men from second class were there onboard?**
- 104
- 108
- 112
- 125

In [12]:
print(data["Pclass"].value_counts())
print(data[data["Sex"] == "male"]["Pclass"].value_counts())
print(data[data["Sex"] == "female"]["Pclass"].value_counts())


Pclass
3    491
1    216
2    184
Name: count, dtype: int64
Pclass
3    347
1    122
2    108
Name: count, dtype: int64
Pclass
3    144
1     94
2     76
Name: count, dtype: int64


In [13]:
print(data["Pclass"].value_counts(normalize=True))
print(data[data["Sex"] == "male"]["Pclass"].value_counts(normalize=True))
print(data[data["Sex"] == "female"]["Pclass"].value_counts(normalize=True))

Pclass
3    0.55
1    0.24
2    0.21
Name: proportion, dtype: float64
Pclass
3    0.60
1    0.21
2    0.19
Name: proportion, dtype: float64
Pclass
3    0.46
1    0.30
2    0.24
Name: proportion, dtype: float64


**3. What are median and standard deviation of `Fare`?. Round to two decimals.**
- median is  14.45, standard deviation is 49.69
- median is 15.1, standard deviation is 12.15
- median is 13.15, standard deviation is 35.3
- median is  17.43, standard deviation is 39.1

In [14]:
print(data["Fare"].median())
print(data["Fare"].std())

14.4542
49.6934285971809


**4. Is that true that the mean age of survived people is higher than that of passengers who eventually died?**
- Yes
- No


In [15]:
print(data[data["Survived"] == 1]["Age"].mean())
print(data[data["Survived"] == 0]["Age"].mean())

28.343689655172415
30.62617924528302


**5. Is that true that passengers younger than 30 y.o. survived more frequently than those older than 60 y.o.? What are shares of survived people among young and old people?**
- 22.7% among young and 40.6% among old
- 40.6% among young and 22.7% among old
- 35.3% among young and 27.4% among old
- 27.4% among young and  35.3% among old

In [ ]:
younger_number = data[(data["Age_category"] == 1) & (data["Survived"] == 1)].shape[0]
print(younger_number)
older_number = data[(data["Age"] > 60) & (data["Survived"] == 1)].shape[0]
print(older_number)
survived_number = data[data["Survived"] == 1].shape[0]
print(younger_number / survived_number, older_number / survived_number)

156
5
0.45614035087719296 0.014619883040935672


In [24]:
young_data = data[data["Age_category"] == 1]
print(young_data["Survived"].mean())
older_number = data[(data["Age"] > 60)]
print(older_number["Survived"].mean())

0.40625
0.22727272727272727


**6. Is that true that women survived more frequently than men? What are shares of survived people among men and women?**
- 30.2% among men and 46.2% among women
- 35.7% among men and 74.2% among women
- 21.1% among men and 46.2% among women
- 18.9% among men and 74.2% among women

In [25]:
male_data = data[data.Sex == "male"]
print(male_data.Survived.mean())
female_data = data[data.Sex == "female"]
print(female_data.Survived.mean())

0.18890814558058924
0.7420382165605095


**7. What's the most popular first name among male passengers?**
- Charles
- Thomas
- William
- John

In [54]:
def get_name(full_name):
    full_name = str(full_name)
    array_from_full_name = full_name.split(' ')
    for i in range(len(array_from_full_name)):
        if array_from_full_name[i] == "Mr.":
            return array_from_full_name[i+1]

    return array_from_full_name[0]


In [55]:
data["Really_name"] = data[(data.Sex == "male") & (data.Sex.notna())]["Name"].apply(get_name)
data["Really_name"].mode()

0    William
Name: Really_name, dtype: object

**8. How is average age for men/women dependent on `Pclass`? Choose all correct statements:**
- On average, men of 1 class are older than 40
- On average, women of 1 class are older than 40
- Men of all classes are on average older than women of the same class
- On average, passengers ofthe first class are older than those of the 2nd class who are older than passengers of the 3rd class

In [63]:
print(data[(data.Sex == "male") & (data.Pclass == 1)]["Age"].median())
print(data[(data.Sex == "female") & (data.Pclass == 1)]["Age"].median())
for pcl in {1, 2, 3}:
    for sex in {"male", "female"}:
        print('Sex: ', sex, "Pclass: ", pcl, "Median age: ", data[(data.Sex == sex) & (data.Pclass == pcl)].Age.median())

40.0
35.0
Sex:  female Pclass:  1 Median age:  35.0
Sex:  male Pclass:  1 Median age:  40.0
Sex:  female Pclass:  2 Median age:  28.0
Sex:  male Pclass:  2 Median age:  30.0
Sex:  female Pclass:  3 Median age:  21.5
Sex:  male Pclass:  3 Median age:  25.0


In [66]:
median_3_class = data[data.Pclass == 3].Age.median()
old_passager_from_2_class = data[data.Age > median_3_class]
print(median_3_class)
print(data[data.Pclass == 1]["Age"].median())
print(old_passager_from_2_class["Age"].median())

24.0
37.0
35.0
